documentation/ADRs/019_log_files_for_model_training.md

```diff
@@ -30,30 +32,68 @@ For related ADRs on the generation of different log files and other general logg
 
 
 ## Decision
-*Detail the decision that was made, including any alternatives that were considered and the reasons for choosing the implemented solution. Provide enough technical specifics to justify the approach.*
+### Logging Content
+- Standard training metrics on W&B, including  training loss, accuracy, epochs, batch size, hyperparameters, etc.
+- Non-standard events -- use `early_terminated` flag on W&B
+  - Early stopping
+  - Diverging loss
+  - Vanishing and exploding gradients
+  - Other anomalies
+
+### Logging Levels
+The logging levels we set in this project are `DEBUG`, `INFO`, `WARNING`, `ERROR`, and `CRITICAL`. The default level is `INFO`, meaning that only levels above will be logged. Below are specific examples for each level that is logged:
+
+#### 1. INFO
+```
+logger.info(f"Training model {config['name']}...")
+```
+At this level logging is used for standard training progress updates.
+
+#### 2. WARNING
+```
+logger.warning(f"DataFrame contains non-np.float64 numeric columns. Converting the following columns: {', '.join(non_float64_cols)}")
+```
+At this level logging is used for minor issues like degraded performance in intermediate steps.
+
+#### 3. ERROR
+```
+logger.error(f"Early stopping at epoch {epoch+1} due to lack of improvement.")
 
-### Overview
-*Overview of the decision in a clear and concise manner.*
+wandb.log({"early_terminated": True})
+```
+At this level logging is used for training stalls, failed checkpoints, or early stops.
+
+#### 4. CRITICAL
+Currently `CRITICAL` hasn't been used, but usually it means major data issues or catastrophic failure requiring immediate attention.
+
+
+### Storage and Distribution:
+The logs are generated at the terminal, in the `.log` file, and on W&B to ensure both internal accessibility and external sharing. Logs on W&B are permanently kept, while the `.log` files are kept until reaching the preset maximum amount .
```

Comment:
Do we know what we are doing with the log files in the long term? Can we store them for longer periods of time to have a historical record maybe? Not sure if this needs a separate ADR or not but something to think about I guess

Reply:
I think it is okay to refer to W&B to check the historical record. The log file at debug and info level doesn't provide enough useful information, and errors are fixed immediately. So I don't think it is super necessary to store them for longer periods.